Customize handlers (#122)
dotnet/docs-maui: docs/user-interface/handlers/customize.md

.NET Multi-platform App UI (.NET MAUI) provides a collection of controls that can be used to display data, initiate actions, indicate activity, display collections, pick data, and more. By default, *handlers* map these cross-platform controls to native controls on each platform. For example, on iOS a .NET MAUI handler maps a .NET MAUI `Button` to an iOS `UIButton`. On Android, the `Button` is mapped to an `AppCompatButton`:

:::image type="content" source="customize-images/button-handler.png" alt-text="Button handler architecture." border="false":::

Handlers can be accessed through a control-specific interface provided by .NET MAUI, such as `IButton` for a `Button`. This avoids the cross-platform control having to reference its handler, and the handler having to reference the cross-platform control. The mapping of the cross-platform control API to the native control API is provided by a *mapper*.

All handler-based .NET MAUI controls can be customized by modifying the mapper for a control's handler. The syntax for modifying a mapper is as follows:

```csharp
HandlerType.MapperName[string] = (handler, view) =>
{
  // Manipulate handler.NativeView to customize the handler
};
```

> All handler types are located in the `Microsoft.Maui.Handlers` namespace, and each handler provides a mapper that maps each cross-platform control property to a method that applies the property value to the native control. The `handler` argument to the mapper provides an instance of the handler being customized, and the `view` argument provides an instance of the cross-platform control that the handler implements. For a list of the type names that implement handler-based .NET MAUI controls, see [Handler-based controls](#handler-based-controls).

Reviewer: I think there was a minor bug here. Perhaps I fixed it?
```suggestion
All handler types are located in the `Microsoft.Maui.Handlers` namespace, and each handler provides a mapper that maps each cross-platform control property to a method that applies the property value to the native control. The `handler` argument provides an instance of the handler being customized, and the `view` argument provides an instance of the cross-platform control that the handler implements. For a list of the type names that implement handler-based .NET MAUI controls, see [Handler-based controls](#handler-based-controls).
```

Each handler class exposes the native control that implements the cross-platform control via its `NativeView` property. This property can be accessed to set native control properties, invoke native control methods, and subscribe to native control events. In addition, the cross-platform control implemented by the handler is exposed via its `VirtualView` property.

> Handlers can be customized to augment the appearance and behavior of a .NET MAUI control beyond the customization that's possible through the control's API. Handlers are global and therefore customization can occur anywhere in your .NET MAUI app. For example, customizing a handler for a control in your `App` class will result in all controls of that type being customized in the app. Similarly, customizing a handler for a control in the first page of your app will also result in all controls of that type in the app being customized.

Reviewer: UPDATE on this comment after reading further:

This suggestion may not make sense. After reading further I think perhaps my understanding about how this all works was wrong and therefore my suggestion is invalid. I'll leave it here though.

----

Maybe be more direct instead of using examples to demonstrate the concept? Maybe you want to tweak this further or ignore it? Just edit my comment here. Is **view** the right term? I'm unsure...
```suggestion
Handlers can be customized to augment the appearance and behavior of a .NET MAUI control beyond the customization that's possible through the control's API. Handler customizations are global and aren't scoped to a specific UI view. Handler customization is allowed to happen anywhere in your app. Once a handler is customized, it affects all native controls of that type, everywhere in your app.
```

Handlers can be customized per platform by using compiler preprocessor directives, to multi-target code based on the platform. Alternatively, you can use partial classes to organize your code into platform-specific folders and files. For more information about conditional compilation, see [Conditional compilation](/dotnet/csharp/language-reference/preprocessor-directives#conditional-compilation).

## Handler lifecycle

All handler-based .NET MAUI controls support two handler lifecycle events:

> - `HandlerChanged` is fired after the handler for a cross-platform control has been created. This event indicates that the native control that implements the cross-platform control is available, and all the property values set on the cross-platform control have been applied to the native control.

Reviewer: You've used `fired` and `fires` a lot 😀 but the proper term for events is `raised`. The code raises the event or the event was raised. Code doesn't fire events and the event wasn't fired.

> - `HandlerChanging` is fired before an existing handler is removed from a cross-platform control, and before a new handler for the cross-platform control is created. This event indicates that the existing native control is about be removed from the cross-platform control, and therefore that any native events should be unwired and other cleanup performed. The `HandlerChangingEventArgs` object that accompanies this event has `OldHandler` and `NewHandler` properties, of type `IElementHandler`.

Reviewer: I would be clear on the `HandlerChanging` and `HandlerChanged` events. to me it sounds like this is true and may be worth calling out (especially if I'm wrong in these assumptions):

01. The `HandlerChanging` event is raised before the `HandlerChanged` event.
01. This doesn't happen the first time handlers are added to controls, since there is nothing being "changed" (removed before an add).

Author: (2) is wrong so I'll clarify.

In addition to these events, each control also has an overridable `OnHandlerChanged` method that's invoked when the `HandlerChanged` event fires, and a `OnHandlerChanging` method that's invoked when the `HandlerChanging` event fires.

## Customize a control with a mapper

> The .NET MAUI `Entry` is a single-line text input control, that implements the `IEntry` interface. On iOS, the `EntryHandler` maps the `Entry` to an iOS `UITextField`. On Android, the `Entry` is mapped to an `AppCompatEditText`, and on Windows the `Entry` is mapped to a `TextBox`:

Reviewer: I think it helps to say control here:
```suggestion
The .NET MAUI `Entry` is a single-line text input control, that implements the `IEntry` interface. On iOS, the `EntryHandler` maps the `Entry` to an iOS `UITextField` control. On Android, the `Entry` is mapped to an `AppCompatEditText`, and on Windows the `Entry` is mapped to a `TextBox` control:
```

:::image type="content" source="customize-images/entry-handler.png" alt-text="Entry handler architecture." border="false":::

The `EntryMapper`, in the `EntryHandler` class, maps the cross-platform control API to the native control API. This mapper can be modified to customize the control on each platform:

```csharp
using Microsoft.Maui;
using Microsoft.Maui.Controls;
using Microsoft.Maui.Graphics;

namespace HandlersDemos.Views
{
    public partial class CustomizeEntryPage : ContentPage
    {
        public CustomizeEntryPage()
        {
            InitializeComponent();

            Microsoft.Maui.Handlers.EntryHandler.EntryMapper["MyCustomization"] = (handler, view) =>
            {
#if __ANDROID__
                handler.NativeView.SetBackgroundColor(Colors.Transparent.ToNative());
#elif __iOS__
                handler.NativeView.BorderStyle = UIKit.UITextBorderStyle.None;
#elif WINDOWS
                handler.NativeView.FontWeight = Microsoft.UI.Text.FontWeights.Thin;
#endif
            };
        }
    }
}
```

> In this example, the `Entry` customization occurs in a page class. Therefore, all `Entry` controls on Android, iOS, and Windows will be customized once the `CustomizeEntryPage` is displayed. The following customization is performed by using compiler preprocessing directives:

Reviewer: Am I correct in this assumption? On the other UI stacks I've used, you have to call Show or add the page to something else. Technically you could do `var page = new CustomizeEntryPage()` and that doesn't show the page, but the constructor has run and now the handler code has run.

```suggestion
In this example, the `Entry` customization occurs in a page class. Therefore, all `Entry` controls on Android, iOS, and Windows will be customized once an instance of the `CustomizeEntryPage` class is created, but before it's displayed. The following customization is performed by using compiler preprocessing directives:
```

> - On Android, the underline is removed from the `Entry`.

Reviewer: Is this correct? The code looks like it's making the background transparent? (unless the background is controlling the visibility of the underline?)

Author: There was an UnderlineVisible property in an earlier preview, which seems to have vanished, so I had to make the background transparent to remove the underline drawable. I'll take another look but this may be as good as it gets now.

Author: Setting the background color to transparent is the way.

- On iOS, the border is removed from the `Entry`.
- On Windows, the thickness of the font in the `Entry` is set to thin.

> [!IMPORTANT]

> > The index into a mapper is an arbitrary string and does not have to correspond to a property exposed by a type, or a property on a native control. For example, `["MyCustomization"]` can be specified as a mapper index, with any native control modification being performed as the customization.

Reviewer: ```suggestion
> The index into a mapper is an arbitrary string and doesn't have to correspond to a property exposed by a type, or a property on a native control. For example, `["MyCustomization"]` can be specified as a mapper index, with any native control modification being performed as the customization.
```

> > The index into a mapper is an arbitrary string and does not have to correspond to a property exposed by a type, or a property on a native control. For example, `["MyCustomization"]` can be specified as a mapper index, with any native control modification being performed as the customization.

Reviewer: Question on this. The "MyCustomization" doesn't make sense in this note and in the previous code example. How does that map to an `Entry` control?

**UPDATE**
Looking at the code below in the next section, there is a type check: `if (view is MyEntry)` so this makes me think that ALL handlers are applied to every control when they're created. There is no system that automatically runs handler X when control X is created. I didn't get that impression until I got this far and the IMPORTANT note confused me and then I saw the following code. That may prove confusing.

Author: I *think* this is resolved from an earlier comment that you went back and edited.

Handlers are global, and customizing a handler for a control will result in all controls of the same type being customized in your app. However, handlers for specific control instances can be customized by subclassing the control, and then by modifying the handler for the base control type only when the control is of the subclassed type. For example, to customize a specific `Entry` control on a page that contains multiple `Entry` controls, you should first subclass the `Entry` control:

```csharp
using Microsoft.Maui.Controls;

namespace HandlersDemos.Controls
{
    public class MyEntry : Entry
    {
    }
}
```

You can then customize the `EntryHandler`, via its mapper, to perform the desired modification to `MyEntry` instances:

```csharp
Microsoft.Maui.Handlers.EntryHandler.EntryMapper["MyCustomization"] = (handler, view) =>
{
    if (view is MyEntry)
    {
#if __ANDROID__
        handler.NativeView.SetBackgroundColor(Colors.Transparent.ToNative());
#elif __iOS__
        handler.NativeView.BorderStyle = UIKit.UITextBorderStyle.None;
#elif WINDOWS
        handler.NativeView.FontWeight = Microsoft.UI.Text.FontWeights.Thin;
#endif
    }
};
```

> Any `MyEntry` instances in the app will then be customized as per the handler modification.

Reviewer: One thing I'm not clear about, do handlers affect already created controls? Or just apply to any control created after the handler is altered?

```csharp
using AndroidX.AppCompat.Widget;
using Microsoft.Maui;
using Microsoft.Maui.Controls;
using Microsoft.Maui.Graphics;
using System;

namespace HandlersDemos.Views
{
    public partial class CustomizeEntryPage : ContentPage
    {
        partial void ChangedHandler(object sender, EventArgs e)
        {
            ((sender as Entry).Handler.NativeView as Android.Views.View).FocusChange += OnFocusChange;
        }

        partial void ChangingHandler(object sender, HandlerChangingEventArgs e)
        {
            if (e.OldHandler != null)
            {
                (e.OldHandler.NativeView as Android.Views.View).FocusChange -= OnFocusChange;
            }
        }

        void OnFocusChange(object sender, EventArgs e)
        {
            var nativeView = sender as AppCompatEditText;

            if (nativeView.IsFocused)
                nativeView.SetBackgroundColor(Colors.LightPink.ToNative());
            else
                nativeView.SetBackgroundColor(Colors.White.ToNative());
        }
    }
}
```

> The advantage of this approach is that compiler preprocessing directives aren't required, and that the partial methods don't have to be implemented on each platform. Instead, the compiler removes references to the method signature on any platforms that don't provide an implementation. For information about partial methods, see [Partial methods](/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods#partial-methods).

Reviewer: > the compiler removes references to the method signature on any platforms that don't provide an implementation.

Is this something special to MAUI? I've never heard of this before. So when compiling for iOS the compiler would remove the `ChangedHandler(sender, e);` call instead of erroring out that a `ChangedHandler` method doesn't exist?

Author: I've re-written this, but yes, there's no error about `ChangedHandler` not existing if it's not implemented on a platform.

The .NET partial method doc says (about partial methods) "If the developer does not supply an implementation, the compiler removes the signature at compile time.".

| `ScrollView` | `IScrollView` | `ScrollViewHandler` | `ScrollViewMapper` |
| `SearchBar` | `ISearchBar` | `SearchBarHandler` | `SearchBarMapper` |
| `ShapeView` | `IShapeView` | `ShapeViewHandler` | `ShapeViewMapper` |
| `Slider` | `ISlider` | `SliderHandler` | `SliderMapper` |
| `Stepper` | `IStepper` | `StepperHandler` | `StepperMapper` |
| `Switch` | `ISwitch` | `SwitchHandler` | `SwitchMapper` |
| `TimePicker` | `ITimePicker` | `TimePickerHandler` | `TimePickerMapper` |
| `View` | `IView` | `ViewHandler` | `ViewMapper` |
| `Window` | `IWindow` | `WindowHandler` | `WindowMapper` |

> [!IMPORTANT]
> The .NET MAUI cross-platform controls are located in the `Microsoft.Maui.Controls` namespace, while handlers for the controls are located in the `Microsoft.Maui.Handlers` namespace. The control interfaces are located in the `Microsoft.Maui` namespace.

## Renderer-based controls

The following controls are backed by renderers, and use a different customization approach:

- `CarouselView`
- `CollectionView`
- `FlyoutPage`
- `IndicatorView`
- `ListView`
- `Map`
- `RefreshView`
- `SwipeView`
- `TabbedPage`
- `TableView`

> For more information about renderers, see [Xamarin.Forms custom renderers](/xamarin/xamarin-forms/app-fundamentals/custom-renderer/).

Reviewer: Xamarin link here?

Author: Removed.
